src/smart-pointers/exercise.rs

```diff
@@ -13,66 +13,81 @@
 // limitations under the License.
 
 // ANCHOR: solution
+use std::cmp::Ordering::*;
+
 // ANCHOR: types
+/// A node in the binary tree.
 #[derive(Debug)]
-struct BinaryTreeNode<T: Ord + Copy> {
+struct Node<T: Ord + Copy> {
     value: T,
-    left: BinaryTree<T>,
-    right: BinaryTree<T>,
+    left: Subtree<T>,
+    right: Subtree<T>,
 }
 
+/// A possibly-empty subtree.
+type Subtree<T> = Option<Box<Node<T>>>;
+
 /// A container storing a set of values, using a binary tree.
 ///
 /// If the same value is added multiple times, it is only stored once.
 #[derive(Debug)]
-pub struct BinaryTree<T: Ord + Copy>(Option<Box<BinaryTreeNode<T>>>);
+pub struct BinaryTree<T: Ord + Copy> {
+    root: Subtree<T>,
+}
 // ANCHOR_END: types
 
 impl<T: Ord + Copy> BinaryTree<T> {
     fn new() -> Self {
-        Self(None)
+        Self { root: None }
     }
 
     fn insert(&mut self, value: T) {
```

Comment:
For each of `insert`, `has`, and `len`, the logic here in `BinaryTree` is reproduced below in `Node`. This definitely feels like the new organization of the types means that the implementation always starts at the "wrong end".